File: steps/math/triangle.py
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
# STEPS - STochastic Engine for Pathway Simulation
# Copyright (C) 2005-2007 Stefan Wils. All rights reserved.
#
# This file is part of STEPS.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301, USA
#
# $Id$
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #


"""A variety of auxiliary functions for dealing with 3D triangles.

The following sets of functionality are offered by this module:
    - Computing triangle areas
    - Barycentric centroids

Currently, all these methods are implemented in Python and we're relying 
on NumPy for speed. If this would ever turn out to be a bottleneck, we'll 
port (parts of) them to C/C++.

SEE ALSO:
    steps.math.tetrahedron
    steps.math.triangle_test
"""


import numpy
import numpy.linalg as linalg


# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #


def area(p, t):
    """Compute the area for one or more triangles.
    
    PARAMETERS:
        p
            An N_pnt * 3 array of points.
        t
            An N_tri * 3 array of integer indices into p.
    
    RETURNS:
        A 1-dimensional array of size N_tri with areas.
    """
    # If we get a 1-dimensional array, make it 2D.
    if len(t.shape) == 1: t = t.reshape(-1,3)
    
    p0 = p[t[:,0],:]
    c = numpy.cross(p[t[:,1],:] - p0, p[t[:,2],:] - p0)
    return 0.5 * numpy.sqrt((c * c).sum(axis = 1))
    

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #


def barycenter(p, t):
    """Compute the barycenter for one or more triangles.
    
    PARAMETERS:
        p
            An N_pnt * 3 array of points.
        t
            An N_tri * 3 array of integer indices into p.
    
    RETURNS:
        A 2-dimensional array of size N_tri * 3 with barycenters.
    """
    return (p[t[:,0],:] + p[t[:,1],:] + p[t[:,2],:]) / 3.0
    
    
# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #


def normal(p, t):
    """Compute the normal vector(s) for one or more triangles.
    
    These normal vectors of course have length 1.0. 
    
    PARAMETERS:
        p
            An N_pnt * 3 array of points.
        t
            An N_tri * 3 array of integer indices into p.
    
    RETURNS:
        A 2-dimensional array of size N_tri * 3 with the normal vectors.
    """
    #ntri = t.shape[0]
    #nrms = numpy.empty((ntri,3))
    #nrms[:,0] = ((p[t[:,1],1] - p[t[:,0],1]) * (p[t[:,2],2] - p[t[:,0],2])) \
    #          - ((p[t[:,1],2] - p[t[:,0],2]) * (p[t[:,2],1] - p[t[:,0],1]))
    #nrms[:,1] = ((p[t[:,1],2] - p[t[:,0],2]) * (p[t[:,2],0] - p[t[:,0],0])) \
    #          - ((p[t[:,1],0] - p[t[:,0],0]) * (p[t[:,2],2] - p[t[:,0],2]))
    #nrms[:,2] = ((p[t[:,1],0] - p[t[:,0],0]) * (p[t[:,2],1] - p[t[:,0],1])) \
    #          - ((p[t[:,1],1] - p[t[:,0],1]) * (p[t[:,2],0] - p[t[:,0],0]))
    #norm = numpy.sqrt((nrms * nrms).sum(axis = 1))
    #norm = numpy.c_ [ norm, norm, norm ]
    #return nrms / norm
    tmp1 = p[t[:,1],:] - p[t[:,0],:]
    tmp2 = p[t[:,2],:] - p[t[:,0],:]
    tmp = numpy.cross(tmp1,tmp2)
    norm = numpy.sqrt((tmp * tmp).sum(axis = 1))
    return tmp / numpy.c_ [ norm, norm, norm ]


# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

# END

File: cpp/math/triangle.cpp
////////////////////////////////////////////////////////////////////////////////
// STEPS - STochastic Engine for Pathway Simulation
// Copyright (C) 2007-2009ÊOkinawa Institute of Science and Technology, Japan.
// Copyright (C) 2003-2006ÊUniversity of Antwerp, Belgium.
//
// See the file AUTHORS for details.
//
// This file is part of STEPS.
//
// STEPSÊis free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// STEPSÊis distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.ÊIf not, see <http://www.gnu.org/licenses/>.
//
////////////////////////////////////////////////////////////////////////////////

/*
 *  Last Changed Rev:  $Rev: 307 $
 *  Last Changed Date: $Date: 2010-03-24 09:25:37 +0900 (Wed, 24 Mar 2010) $
 *  Last Changed By:   $Author: wchen $
 */

// STL headers.
#include <cassert>
#include <cmath>

// STEPS headers.
#include "../common.h"
#include "triangle.hpp"

////////////////////////////////////////////////////////////////////////////////

static inline void cross(double * v0, double * v1, double * vo)
{
    vo[0] = (v0[1] * v1[2]) - (v0[2] * v1[1]);
    vo[1] = (v0[2] * v1[0]) - (v0[0] * v1[2]);
    vo[2] = (v0[0] * v1[1]) - (v0[1] * v1[0]);
}

////////////////////////////////////////////////////////////////////////////////

double steps::math::triArea(double * v0, double * v1, double * v2)
{
    double vv[3];
    vv[0] = v1[0] - v0[0];
    vv[1] = v1[1] - v0[1];
    vv[2] = v1[2] - v0[2];
    double ww[3];
    ww[0] = v2[0] - v0[0];
    ww[1] = v2[1] - v0[1];
    ww[2] = v2[2] - v0[2];
    double cc[3];
    cross(vv, ww, cc);
    return 0.5 * sqrt((cc[0] * cc[0]) + (cc[1] * cc[1]) + (cc[2] * cc[2]));
}

////////////////////////////////////////////////////////////////////////////////

void steps::math::triBarycenter
(
    double * v0, double * v1, double * v2,
    double * po
)
{
    po[0] = (v0[0] + v1[0] + v2[0]) / 3.0;
    po[1] = (v0[1] + v1[1] + v2[1]) / 3.0;
    po[2] = (v0[2] + v1[2] + v2[2]) / 3.0;
}

////////////////////////////////////////////////////////////////////////////////

void steps::math::triNormal
(
    double * v0, double * v1, double * v2,
    double * vo
)
{
    double vv[3];
    vv[0] = v1[0] - v0[0];
    vv[1] = v1[1] - v0[1];
    vv[2] = v1[2] - v0[2];
    double ww[3];
    ww[0] = v2[0] - v0[0];
    ww[1] = v2[1] - v0[1];
    ww[2] = v2[2] - v0[2];
    double cc[3];
    cross(vv, ww, cc);
    double norm = sqrt((cc[0] * cc[0]) + (cc[1] * cc[1]) + (cc[2] * cc[2]));
    vo[0] = cc[0] / norm;
    vo[1] = cc[1] / norm;
    vo[2] = cc[2] / norm;
}

////////////////////////////////////////////////////////////////////////////////

// END
